### DISCUSSION TITLE
How to get the workflow activities already sorted

### DISCUSSION THREAD

**Discussion by landonzeng** [elsa-workflows/elsa-core#2455] [Ideas]
How to get the workflow activities that have been sorted, for example: I now need to get all the nodes of the workflow, now I have got the json data of the workflow, but the order of them is not sorted, is there an interface to get the data that has been sorted

`{
  "$id": "1",
  "activities": [
    {
      "$id": "2",
      "activityId": "30b9eace-1cbd-422f-a371-9f31460c86b2",
      "type": "SendHttpRequest",
      "name": "activity1",
      "displayName": "Send HTTP Request",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "3",
          "name": "Url",
          "expressions": {
            "$id": "4",
            "Literal": "http://47.103.96.35:8091/api/Account/Get?id=12"
          }
        },
        {
          "$id": "5",
          "name": "Method",
          "expressions": {
            "$id": "6"
          }
        },
        {
          "$id": "7",
          "name": "Content",
          "expressions": {
            "$id": "8"
          }
        },
        {
          "$id": "9",
          "name": "ContentType",
          "expressions": {
            "$id": "10",
            "Literal": "application/json"
          }
        },
        {
          "$id": "11",
          "name": "ReadContent",
          "expressions": {
            "$id": "12",
            "Literal": "true"
          }
        },
        {
          "$id": "13",
          "name": "SupportedStatusCodes",
          "expressions": {
            "$id": "14"
          }
        },
        {
          "$id": "15",
          "name": "Authorization",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "16",
            "JavaScript": "return input.Headers.Authorization;"
          }
        },
        {
          "$id": "17",
          "name": "RequestHeaders",
          "expressions": {
            "$id": "18"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "19",
      "activityId": "17fa9d04-fb95-4007-b9ff-eae7ecaf98f3",
      "type": "WriteHttpResponse",
      "displayName": "HTTP Response",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "20",
          "name": "Content",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "21",
            "JavaScript": "return JSON.stringify(activities.activity1.ResponseContent());"
          }
        },
        {
          "$id": "22",
          "name": "ContentType",
          "expressions": {
            "$id": "23",
            "Literal": "application/json"
          }
        },
        {
          "$id": "24",
          "name": "StatusCode",
          "expressions": {
            "$id": "25"
          }
        },
        {
          "$id": "26",
          "name": "CharSet",
          "expressions": {
            "$id": "27"
          }
        },
        {
          "$id": "28",
          "name": "ResponseHeaders",
          "expressions": {
            "$id": "29"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "30",
      "activityId": "fb88fde8-ab75-4654-95b3-d63cb0f47c2b",
      "type": "HttpEndpoint",
      "displayName": "HTTP Endpoint",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "31",
          "name": "Path",
          "expressions": {
            "$id": "32",
            "Literal": "/api/test"
          }
        },
        {
          "$id": "33",
          "name": "Methods",
          "expressions": {
            "$id": "34",
            "Json": "[\"GET\"]"
          }
        },
        {
          "$id": "35",
          "name": "ReadContent",
          "expressions": {
            "$id": "36",
            "Literal": "true"
          }
        },
        {
          "$id": "37",
          "name": "TargetType",
          "expressions": {
            "$id": "38"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "39",
      "activityId": "86fc9a88-ee50-4700-bb96-1620d61e8d07",
      "type": "If",
      "displayName": "If/Else",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "40",
          "name": "Condition",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "41",
            "JavaScript": "if(activities.activity1.ResponseContent.status==0){\r\n    return true;\r\n}else{\r\n    return false;\r\n}"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "42",
      "activityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
      "type": "If",
      "displayName": "If/Else",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "43",
          "name": "Condition",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "44",
            "JavaScript": "return activities.ApproveActivity.SignalInput().isTrue;"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "45",
      "activityId": "3e8ff4dc-b8db-4d91-b9d2-f73eaa994a69",
      "type": "WriteLine",
      "displayName": "Write Line",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "46",
          "name": "Text",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "47",
            "JavaScript": "return activities.ApproveActivity.Output();",
            "Liquid": "{{Input|json}}"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "48",
      "activityId": "5638827f-8f68-4558-9d30-30422dcd94bf",
      "type": "WriteLine",
      "displayName": "Write Line",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "49",
          "name": "Text",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "50",
            "JavaScript": "return activities.ApproveActivity.Output();",
            "Liquid": "return activities.ApproveActivity.Output();"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "51",
      "activityId": "0d9f5e40-bf1e-4883-83c3-0baf9d1075ff",
      "type": "SignalReceived",
      "name": "ApproveActivity",
      "displayName": "ApproveActivity",
      "description": "ApproveActivity",
      "persistWorkflow": true,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "52",
          "name": "Signal",
          "expressions": {
            "$id": "53",
            "Literal": "Approve"
          }
        }
      ],
      "propertyStorageProviders": {}
    },
    {
      "$id": "54",
      "activityId": "84dc44cf-c563-4a0d-8b76-d0c1d5a29b16",
      "type": "WriteLine",
      "displayName": "Write Line",
      "persistWorkflow": false,
      "loadWorkflowContext": false,
      "saveWorkflowContext": false,
      "properties": [
        {
          "$id": "55",
          "name": "Text",
          "syntax": "JavaScript",
          "expressions": {
            "$id": "56",
            "JavaScript": "return activities.ApproveActivity.Output();"
          }
        }
      ],
      "propertyStorageProviders": {}
    }
  ],
  "connections": [
    {
      "$id": "57",
      "sourceActivityId": "30b9eace-1cbd-422f-a371-9f31460c86b2",
      "NextId": "17fa9d04-fb95-4007-b9ff-eae7ecaf98f3",
      "outcome": "Done"
    },
    {
      "$id": "58",
      "sourceActivityId": "fb88fde8-ab75-4654-95b3-d63cb0f47c2b",
      "targetActivityId": "30b9eace-1cbd-422f-a371-9f31460c86b2",
      "outcome": "Done"
    },
    {
      "$id": "59",
      "sourceActivityId": "17fa9d04-fb95-4007-b9ff-eae7ecaf98f3",
      "targetActivityId": "86fc9a88-ee50-4700-bb96-1620d61e8d07",
      "outcome": "Done"
    },
    {
      "$id": "60",
      "sourceActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
      "targetActivityId": "3e8ff4dc-b8db-4d91-b9d2-f73eaa994a69",
      "outcome": "True"
    },
    {
      "$id": "61",
      "sourceActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
      "targetActivityId": "5638827f-8f68-4558-9d30-30422dcd94bf",
      "outcome": "False"
    },
    {
      "$id": "62",
      "sourceActivityId": "86fc9a88-ee50-4700-bb96-1620d61e8d07",
      "targetActivityId": "0d9f5e40-bf1e-4883-83c3-0baf9d1075ff",
      "outcome": "False"
    },
    {
      "$id": "63",
      "sourceActivityId": "0d9f5e40-bf1e-4883-83c3-0baf9d1075ff",
      "targetActivityId": "84dc44cf-c563-4a0d-8b76-d0c1d5a29b16",
      "outcome": "Done"
    },
    {
      "$id": "64",
      "sourceActivityId": "84dc44cf-c563-4a0d-8b76-d0c1d5a29b16",
      "targetActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
      "outcome": "Done"
    }
  ],
  "variables": {
    "$id": "65",
    "data": {}
  },
  "customAttributes": {
    "$id": "66",
    "data": {}
  }
}`

**Comment by landonzeng** [comment]
The data of the connections should be used for sorting, is there any current method to sort and output?
`[
  {
    "$id": "57",
    "sourceActivityId": "30b9eace-1cbd-422f-a371-9f31460c86b2",
    "targetActivityId": "17fa9d04-fb95-4007-b9ff-eae7ecaf98f3",
    "outcome": "Done"
  },
  {
    "$id": "58",
    "sourceActivityId": "fb88fde8-ab75-4654-95b3-d63cb0f47c2b",
    "targetActivityId": "30b9eace-1cbd-422f-a371-9f31460c86b2",
    "outcome": "Done"
  },
  {
    "$id": "59",
    "sourceActivityId": "17fa9d04-fb95-4007-b9ff-eae7ecaf98f3",
    "targetActivityId": "86fc9a88-ee50-4700-bb96-1620d61e8d07",
    "outcome": "Done"
  },
  {
    "$id": "60",
    "sourceActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
    "targetActivityId": "3e8ff4dc-b8db-4d91-b9d2-f73eaa994a69",
    "outcome": "True"
  },
  {
    "$id": "61",
    "sourceActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
    "targetActivityId": "5638827f-8f68-4558-9d30-30422dcd94bf",
    "outcome": "False"
  },
  {
    "$id": "62",
    "sourceActivityId": "86fc9a88-ee50-4700-bb96-1620d61e8d07",
    "targetActivityId": "0d9f5e40-bf1e-4883-83c3-0baf9d1075ff",
    "outcome": "False"
  },
  {
    "$id": "63",
    "sourceActivityId": "0d9f5e40-bf1e-4883-83c3-0baf9d1075ff",
    "targetActivityId": "84dc44cf-c563-4a0d-8b76-d0c1d5a29b16",
    "outcome": "Done"
  },
  {
    "$id": "64",
    "sourceActivityId": "84dc44cf-c563-4a0d-8b76-d0c1d5a29b16",
    "targetActivityId": "dfa26f80-653a-4f58-b44e-4043f24ef0e4",
    "outcome": "Done"
  }
]`

**Reply by sfmskywalker** [reply]
Hi @landonzeng ,

Kindly open separate discussions for different questions. This makes it easier for participants to discuss & answer the question 😄 

Thanks for understanding.

**Reply by landonzeng** [reply]
Sorry, I am not too skilled in using github, I will pay attention in the future

**Reply by sfmskywalker** [reply]
No problem, thank you for understanding.

**Comment by sfmskywalker** [comment]
There's currently no option to request a sorted list of activity types, but we can add it if you want. If so, please open an issue that we can track.
